src/core/send.c

```diff
@@ -1218,6 +1247,24 @@ QuicSendFlush(
         // operation is queued to send the rest.
         //
         QuicSendQueueFlush(&Connection->Send, REASON_SCHEDULING);
+
+        if (Builder.TotalCountDatagrams + 1 > Connection->PeerPacketTolerance) {
+            //
+            // We're scheduling limited, so we should tell the peer to use our
+            // (max) batch size + 1 as the peer tolerance as a hint that they
+            // should expect more than a single batch before needing to send an
+            // acknowledgement back.
+            //
+            QuicConnUpdatePeerPacketTolerance(Connection, Builder.TotalCountDatagrams + 1);
+        }
+
+    } else if (Builder.TotalCountDatagrams > Connection->PeerPacketTolerance) {
```

Comment:
Can Builder.TotalCountDatagrams change frequently across sends?

Reply:
It's a function of congestion window, pacing and application payload size. So yes, I'd say.

Reply:
In that situation, you'd be sending a lot of these ackfreq frames, right?

Reply:
Well, since we only send if this increases (unless there's persistent congestion) I don't expect that many. But that being said, if we happen to send 10 or 15 is that awful?

Reply:
Ah! I missed that this was a watermark. So then, how about issues from packet tolerance being too large?

Reply:
There is always the delayed ACK timer. This is just increasing the threshold for immediate ACKs. And the final safety net is reset on persistent congestion. But I don't have any "right" answer here. The goal is to reduce the CPU costs of ACKs in the really high throughput scenarios. That CPU usage is the last known QUIC-specific logic bottleneck. I'm just experimenting here really.

Reply:
Right, I was going to ask about that too: what's the connection between persistent congestion and the packet tolerance?

Reply:
The idea here is to inform the peer of the send batch size (indirectly) so they ack at a similar batch size. If we hit persistent congestion, we're going to end up resetting our CWND which will result in a reset of the send batch size. Additionally, if we're seeing that much loss, we want to make sure the peer is ACK'ing more frequently to account for loss.

Reply:
Got it.